### PR TITLE
Big gulpfile refactor

### PR DESCRIPTION
**Pull request by robdodson**
@addyosmani @chuckh @abdonrd @FredKSchott @justinfagnani

This is my attempt to move a lot of the polymer-build boilerplate into its own gulp task so folks don't have to look at it or be confused by it.

This also uses gulp 4 cuz is cool.

PTAL!

### CODE REVIEW

**Comment by addyosmani** [comment]
Reviewed in person with Rob:
- I found this approach easier to read and reason about than the existing gulpfile
- It was clear where I might inject my own source processes ([here](https://github.com/PolymerElements/generator-polymer-init-custom-build/pull/14/commits/68c9ac1a6b8461db5ad720a4688565061bc1cdd8#diff-e176445f4818b41bb3c262d20eefc784R50)) or tasks (gulp-tasks directory)
- Preserves the ability to go back and edit most of the core logic that currently exists, shifting it away from the main developer entry point over into the [tasks directory](https://github.com/PolymerElements/generator-polymer-init-custom-build/pull/14/commits/68c9ac1a6b8461db5ad720a4688565061bc1cdd8#diff-d59df940072c02caddf2cb55131c5c31R2)

The only nits I have here are related to SW configuration and are present in the current master too. Although polymerbuild uses sw-precache, we could probably do a better job of just reminding folks that config information for that can be found over in the precache repo and link it up in a comment. If I was looking at project.js for the first time I'd probably otherwise get a little confused.

There were also one or two spots during the review with Rob where we didn't know what everything in the pipeline did (analyzing for example). Comments around those (if we need to expand further) could help. 

Other than that, all looks good to me 👍

File: generators/app/gulp-tasks/project.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    ]);
+  }
+}
+
+// Generates a service worker for the bundled and unbundled version of the site
+// using the settings from the global.config
+function serviceWorker() {
+  const swConfig = {
+    navigateFallback: global.config.fallback,
+  };
```
Comment (abdonrd) [inline generators/app/gulp-tasks/project.js]: @robdodson @chuckh @FredKSchott @addyosmani maybe we need to get this info from the `polymer.json` or from the `sw-precache-config.js`.

File: generators/app/gulp-tasks/project.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+'use strict';
+
+const mergeStream = require('merge-stream');
+const gulp = require('gulp');
+
+const polymer = require('polymer-build');
+const PolymerProject = polymer.PolymerProject;
+const fork = polymer.forkStream;
+const addServiceWorker = polymer.addServiceWorker;
```
Comment (abdonrd) [inline generators/app/gulp-tasks/project.js]: @robdodson I would eliminate these three `polymer.something` variables. I would use them directly.

File: generators/app/gulpfile.js [hunk 3]
```diff
@@ -10,90 +10,64 @@
...
+
+// Got problems? Try logging 'em
+// logging.setVerbose();
+
+// !!! IMPORTANT !!! //
+// Keep the global.config above any of the gulp-tasks that depend on it
+global.config = {
+  json: path.join(process.cwd(), 'polymer.json'),
+  env: 'prod',
+  src: 'src',
```
Comment (abdonrd) [inline generators/app/gulpfile.js]: @robdodson you don't use `prod` and `src`.

File: generators/app/gulpfile.js [hunk 4]
```diff
@@ -10,90 +10,64 @@
...
+const gulpif = require('gulp-if');
+const logging = require('plylog');
+
+// Got problems? Try logging 'em
+// logging.setVerbose();
+
+// !!! IMPORTANT !!! //
+// Keep the global.config above any of the gulp-tasks that depend on it
+global.config = {
+  json: path.join(process.cwd(), 'polymer.json'),
```
Comment (abdonrd) [inline generators/app/gulpfile.js]: Rename to `polymerJsonPath`?

File: generators/app/gulpfile.js [hunk 5]
```diff
@@ -10,90 +10,64 @@
...
+  env: 'prod',
+  src: 'src',
+  dest: 'build',
```
Comment (abdonrd) [inline generators/app/gulpfile.js]: Rename to `buildPath`?

File: generators/app/gulpfile.js [hunk 6]
```diff
@@ -10,90 +10,64 @@
...
+  bundled: 'build/bundled',
+  unbundled: 'build/unbundled',
+  serviceWorkerPath: 'service-worker.js',
+  fallback: '/index.html'
```
Comment (abdonrd) [inline generators/app/gulpfile.js]: Rename to `serviceWorkerNavigateFallback`?

File: generators/app/gulpfile.js [hunk 7]
```diff
@@ -10,90 +10,64 @@
  * Based on: https://github.com/Polymer/polymer-build/blob/master/test/test-project/gulpfile.js
```
Comment (abdonrd) [inline generators/app/gulpfile.js]: We probably remove this line? There are a lot of changes.

File: generators/app/gulp-tasks/project.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      serviceWorkerPath: global.config.serviceWorkerPath,
+    }),
+    // Once the bundled build stream is complete, create a service worker for
+    // the build
+    addServiceWorker({
+      project: project,
+      buildRoot: global.config.bundled,
+      swConfig: swConfig,
+      serviceWorkerPath: global.config.serviceWorkerPath,
+      bundled: true,
```
Comment (abdonrd) [inline generators/app/gulp-tasks/project.js]: Remove `,`

File: generators/app/gulp-tasks/project.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    })
+  ])
```
Comment (abdonrd) [inline generators/app/gulp-tasks/project.js]: Missed `;`

**Comment by abdonrd** [comment]
I create a PR fixing my comments: https://github.com/PolymerElements/generator-polymer-init-custom-build/pull/15

**Comment by abdonrd** [comment]
Another thing... I get this when run gulp:

screen shot 2016-08-11 at 09 56 54

Maybe we want to remove `'bundle'` task and add `build-bundle` and `build-unbundle`.

A lot of people only want the bundled version for now.

**Comment by robdodson** [comment]
@addyosmani 

> There were also one or two spots during the review with Rob where we didn't know what everything in the pipeline did (analyzing for example). Comments around those (if we need to expand further) could help.

I'd like to link to the polymer-build README for these. Waiting on https://github.com/Polymer/polymer-build/pull/19 to be merged

File: generators/app/gulpfile.js (hunk 3 above)
Comment (robdodson) [inline generators/app/gulpfile.js]: yeah true. I noticed the `wsk-tasks` use the `env` var to do different things and was thinking about maybe stealing some of their examples...

**Comment by chuckh** [comment]
I agree with @abdonrd about `bundle` and `unbunbled`.  We could do @abdonrd and @FredKSchott suggested syntax:

- `gulp build`: `build:bundled` & `build:unbundled`
- `gulp build:bundled`: only build bundled
- `gulp build:unbundled`: only build unbundled

**Comment by abdonrd** [comment]
Hi @googlebot!

File: generators/app/package.json [hunk 10]
```diff
@@ -16,7 +16,7 @@
   "private": true,
   "devDependencies": {
     "del": "^2.2.1",
-    "gulp": "^3.9.1",
+    "gulp": "gulpjs/gulp#4.0",
```
Comment (chuckh) [inline generators/app/package.json]: Will moving to gulp 4 force users of custom-build to update all their other gulp files in all other projects that are still in gulp 3 syntax?

File: generators/app/gulp-tasks/project.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      swConfig: swConfig,
+      serviceWorkerPath: global.config.serviceWorker.filename,
+    }),
+    // Once the bundled build stream is complete, create a service worker for
+    // the build
+    polymer.addServiceWorker({
+      project: project,
+      buildRoot: bundledPath,
+      swConfig: swConfig,
+      serviceWorkerPath: global.config.serviceWorker.filename,
```
Comment (chuckh) [inline generators/app/gulp-tasks/project.js]: remove comma.

File: generators/app/gulpfile.js [hunk 12]
```diff
@@ -6,94 +6,68 @@
...
-   });
- });
-
- gulp.task('default', ['service-worker']);
+'use strict';
+
+const path = require('path');
+const gulp = require('gulp');
+const gulpif = require('gulp-if');
+const logging = require('plylog');
```
Comment (chuckh) [inline generators/app/gulpfile.js]: Consider moving `const logging = require('plylog');` to line above `// logging.setVerbose();` and comment it out.

**Comment by chuckh** [comment]
Current Eslinting errors we should consider fixing.

```
/Users/chuckh/DevPolymerLocal/custom-build/generator-polymer-init-custom-build/generators/app/gulpfile.js
  16:7  error  'logging' is defined but never used  no-unused-vars

/Users/chuckh/DevPolymerLocal/custom-build/generator-polymer-init-custom-build/generators/app/index.js
  15:7   error  'chalk' is defined but never used          no-unused-vars
  17:7   error  'yosay' is defined but never used          no-unused-vars
  20:20  error  Missing space before function parentheses  space-before-function-paren
  30:54  error  Unexpected trailing comma                  comma-dangle
  34:20  error  Missing space before function parentheses  space-before-function-paren

/Users/chuckh/DevPolymerLocal/custom-build/generator-polymer-init-custom-build/generators/app/gulp-tasks/images.js
  17:19  error  There should be no space after '{'   object-curly-spacing
  17:57  error  There should be no space before '}'  object-curly-spacing

/Users/chuckh/DevPolymerLocal/custom-build/generator-polymer-init-custom-build/generators/app/gulp-tasks/project.js
  60:19  error  Unexpected parentheses around single function argument  arrow-parens
  68:19  error  Unexpected parentheses around single function argument  arrow-parens
  82:67  error  Unexpected trailing comma                               comma-dangle
  92:62  error  Unexpected trailing comma                               comma-dangle
```

**Comment by robdodson** [comment]
@chuckh let's do the linting errors in a different PR. We may also need to change our lint settings.

File: generators/app/gulpfile.js (hunk 12 above)
Comment (robdodson) [inline generators/app/gulpfile.js]: good idea

File: generators/app/package.json (hunk 10 above)
Comment (robdodson) [inline generators/app/package.json]: nope. I think the gulp 4 dependency is local to the project

**Comment by abdonrd** [comment]
LGTM!

**Comment by chuckh** [comment]
LGTM2! Nice work @robdodson @abdonrd.

**Comment by abdonrd** [comment]
Oops! Error if I select `bundled` instead of `both` in the `global.config.build.bundleType`:

screen shot 2016-08-12 at 01 44 46

**Comment by abdonrd** [comment]
@robdodson fixed in https://github.com/PolymerElements/generator-polymer-init-custom-build/pull/16

**Pull request merged**